In [1]:
import os

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
os.chdir('..')
path = os.path.join(os.getcwd(),'Datasets\\creditcard.csv')
path

'D:\\Babin\\Internship\\Fusemachine-Internship\\Data Science\\Datasets\\creditcard.csv'

# Handling Imbalanced Dataset with Machine Learning

In [4]:
df = pd.read_csv(path)
df

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

### Checking if there are null values

In [5]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [6]:
df.shape

(284807, 31)

In [7]:
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

Here, we can see that there are 284315 samples for the target label 0 but there are only 492 samples for the target label 1. Hence, we can say that the dataset is imbalanced.

In [8]:
X = df.drop(['Class'],axis=1) #Independent Features
y = df['Class']  #Dependent Features

## Cross Validation and Hyperparameter Tuning

In [9]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
from sklearn.model_selection import KFold
import numpy as np
from sklearn.model_selection import GridSearchCV

In [10]:
from sklearn.model_selection import train_test_split

In [11]:
X_train,X_test,y_train,y_test = train_test_split(X,y,train_size=0.7)

In [12]:
10.0**np.arange(-2,3)

array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02])

In [13]:
log_class = LogisticRegression()
grid = {
 'C': 10.0**np.arange(-2,3),
 'penalty': ['l1','l2']
}
cv = KFold(n_splits=5,shuffle=False,random_state = None)

In [14]:
clf = GridSearchCV(log_class,grid,cv=cv,n_jobs=-1,scoring='f1_macro')

In [15]:
clf.fit(X_train,y_train)

C:\Users\Predator\Envs\kn_ml\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
25 fits failed out of a total of 50.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
25 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Predator\Envs\kn_ml\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Predator\Envs\kn_ml\lib\site-packages\sklearn\linear_model\_logistic.py", line 1461, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "C:\Users\Predator\Envs\kn_ml\lib\site-packages\sklearn\linear_model\_logistic.py", line 447, in _check_solver
    raise 

GridSearchCV(cv=KFold(n_splits=5, random_state=None, shuffle=False),
             estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02]),
                         'penalty': ['l1', 'l2']},
             scoring='f1_macro')

In [16]:
y_pred = clf.predict(X_test)
y_pred

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [17]:
print(confusion_matrix(y_test,y_pred))

[[85263    38]
 [   51    91]]


In [18]:
print(accuracy_score(y_test,y_pred))

0.9989583699074237


In [19]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85301
           1       0.71      0.64      0.67       142

    accuracy                           1.00     85443
   macro avg       0.85      0.82      0.84     85443
weighted avg       1.00      1.00      1.00     85443



In [20]:
from sklearn.ensemble import RandomForestClassifier

In [21]:
classifier = RandomForestClassifier()
classifier.fit(X_train,y_train)

RandomForestClassifier()

In [22]:
y_pred = classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[85291    10]
 [   35   107]]
0.9994733330992591
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85301
           1       0.91      0.75      0.83       142

    accuracy                           1.00     85443
   macro avg       0.96      0.88      0.91     85443
weighted avg       1.00      1.00      1.00     85443



In [23]:
class_weight = dict({0:1,1:100})
class_weight

{0: 1, 1: 100}

In [24]:
classifier_ww = RandomForestClassifier(class_weight=class_weight)
classifier_ww.fit(X_train,y_train)

RandomForestClassifier(class_weight={0: 1, 1: 100})

In [26]:
y_pred2 = classifier_ww.predict(X_test)
print(confusion_matrix(y_test,y_pred2))
print(accuracy_score(y_test,y_pred2))
print(classification_report(y_test,y_pred2))

[[85292     9]
 [   38   104]]
0.9994499256814484
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85301
           1       0.92      0.73      0.82       142

    accuracy                           1.00     85443
   macro avg       0.96      0.87      0.91     85443
weighted avg       1.00      1.00      1.00     85443



## Undersampling

Here, we try to reduce the points of the maximum labels.
Suppose, we have two labels 0 and 1 and we have 10000 0 labels and only 100 1 labels. Then we try to reduce the number of samples for 0.

In [27]:
y_train.shape

(199364,)

In [28]:
y_train.value_counts()

0    199014
1       350
Name: Class, dtype: int64

In [29]:
from collections import Counter
Counter(y_train)

Counter({0: 199014, 1: 350})

In [30]:
from imblearn.under_sampling import NearMiss
ns = NearMiss(0.8)
X_train_ns,y_train_ns = ns.fit_resample(X_train,y_train)
print(f'The number of classes before fit {Counter(y_train)}')
print(f'The number of classes after fit {Counter(y_train_ns)}')

C:\Users\Predator\Envs\kn_ml\lib\site-packages\imblearn\utils\_validation.py:586: FutureWarning: Pass sampling_strategy=0.8 as keyword args. From version 0.9 passing these as positional arguments will result in an error
  warnings.warn(


The number of classes before fit Counter({0: 199014, 1: 350})
The number of classes after fit Counter({0: 437, 1: 350})


In [34]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier()
classifier.fit(X_train_ns,y_train_ns)

RandomForestClassifier()

In [35]:
y_pred2 = classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred2))
print(accuracy_score(y_test,y_pred2))
print(classification_report(y_test,y_pred2))

[[58784 26517]
 [    9   133]]
0.6895474175766301
              precision    recall  f1-score   support

           0       1.00      0.69      0.82     85301
           1       0.00      0.94      0.01       142

    accuracy                           0.69     85443
   macro avg       0.50      0.81      0.41     85443
weighted avg       1.00      0.69      0.81     85443



## Over Sampling

Here, we try to increase the samples of the minimum labels.
Suppose, we have two labels 0 and 1 and we have 10000 0 labels and only 100 1 labels. Then we try to increase the number of samples for 1 by picking samples at random with replacement.

In [36]:
from imblearn.over_sampling import RandomOverSampler

In [37]:
os = RandomOverSampler(0.5)
X_train_os, y_train_os = os.fit_resample(X_train,y_train)
print(f'The number of classes before fit {Counter(y_train)}')
print(f'The number of classes after fit {Counter(y_train_os)}')

C:\Users\Predator\Envs\kn_ml\lib\site-packages\imblearn\utils\_validation.py:586: FutureWarning: Pass sampling_strategy=0.5 as keyword args. From version 0.9 passing these as positional arguments will result in an error
  warnings.warn(


The number of classes before fit Counter({0: 199014, 1: 350})
The number of classes after fit Counter({0: 199014, 1: 99507})


In [38]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier()
classifier.fit(X_train_os,y_train_os)

y_pred2 = classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred2))
print(accuracy_score(y_test,y_pred2))
print(classification_report(y_test,y_pred2))

[[85292     9]
 [   31   111]]
0.9995318516437859
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85301
           1       0.93      0.78      0.85       142

    accuracy                           1.00     85443
   macro avg       0.96      0.89      0.92     85443
weighted avg       1.00      1.00      1.00     85443



## SMOTETokmek - Synthetic Minority Over-sampling Technique 
- Combine over- and under-sampling using SMOTE and Tomek links.

In [42]:
# from imblearn.combine import SMOTETomek

# st = SMOTETomek(random_state = 42)
# X_train_st, y_train_st = st.fit_resample(X_train,y_train)
# print(f'The number of classes before fit {Counter(y_train)}')
# print(f'The number of classes after fit {Counter(y_train_st)}')

In [43]:
# from sklearn.ensemble import RandomForestClassifier
# classifier = RandomForestClassifier()
# classifier.fit(X_train_st,y_train_st)

# y_pred2 = classifier.predict(X_test)
# print(confusion_matrix(y_test,y_pred2))
# print(accuracy_score(y_test,y_pred2))
# print(classification_report(y_test,y_pred2))

## Ensemple Techniques

In [44]:
from imblearn.ensemble import EasyEnsembleClassifier

In [46]:
easy = EasyEnsembleClassifier()

In [47]:
easy.fit(X_train,y_train)

EasyEnsembleClassifier()

In [48]:
y_pred = easy.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[81649  3652]
 [    9   133]]
0.957152721697506
              precision    recall  f1-score   support

           0       1.00      0.96      0.98     85301
           1       0.04      0.94      0.07       142

    accuracy                           0.96     85443
   macro avg       0.52      0.95      0.52     85443
weighted avg       1.00      0.96      0.98     85443

